File: src/gha/Components/ShatterCrossingsComponent.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using GeomSeq.Native;
using Grasshopper.Kernel;
using Grasshopper.Kernel.Data;
using Grasshopper.Kernel.Types;
using Rhino.Geometry;

namespace GeomSeq.Components;

public sealed class ShatterCrossingsComponent : GH_Component
{
    // Contract: never change. Saved .gh files find this component by it.
    private static readonly Guid Id = new("3805baf9-7e11-4aff-b8bb-2a1c63881800");

    // Pieces are rebuilt from the vertices they came from, so an uncut segment
    // reproduces its ends exactly and a shared vertex is bit-identical. This only
    // has to absorb the plane round-trip, which both sides of a joint go through
    // together -- it is not a modelling tolerance, and Tolerance is not it.
    private const double JoinTol = 1e-9;

    public ShatterCrossingsComponent()
        : base("Shatter Crossings", "ShatterX",
               "Cuts polylines where they meet and opens a gap there, so crossing paths no longer touch. " +
               "Covers a T-junction as well as a crossing. For a meeting pair the earlier input is left whole " +
               "and the later one yields, so reorder the input to choose which path survives intact.",
               "GeomSeq", "Division")
    {
    }

    public override Guid ComponentGuid => Id;
    public override GH_Exposure Exposure => GH_Exposure.primary;
    protected override Bitmap Icon => Icons.ShatterCrossings;

    // Contract: Grasshopper saves wires by port index. New ports go at the end, never in between.
    //
    // Curves is a tree rather than a list for the same reason Redistribute Lookups takes
    // one: under list access Grasshopper solves once per branch and appends the iteration
    // index to every output path, which would land the results one level deeper than the
    // input index this component exists to preserve.
    protected override void RegisterInputParams(GH_InputParamManager p)
    {
        p.AddCurveParameter("Curves", "C",
            "Polylines to shatter. Each branch is handled on its own, so curves in different branches never cut each other.",
            GH_ParamAccess.tree);
        p.AddNumberParameter("Distance", "D",
            "The whole gap: how far apart the two cut ends end up. Half of it comes off either side of each contact.",
            GH_ParamAccess.item, 0.0);
        p.AddNumberParameter("Tolerance", "T",
            "How close counts as touching rather than missing, in model units. A T-junction is rarely drawn exact, " +
            "so the document tolerance is usually right. 0 means exact only.",
            GH_ParamAccess.item, 0.0);
        p.AddBooleanParameter("Self", "X",
            "Also cut a polyline where it crosses itself. Its own corners are never cut either way.",
            GH_ParamAccess.item, false);
        p.AddPlaneParameter("Plane", "P",
            "Plane the contacts are measured in. Height above it rides along, so a cut point keeps the height it would have had.",
            GH_ParamAccess.item, Plane.WorldXY);

        // Optional so empty input reaches SolveInstance and gets a Remark rather than
        // Grasshopper's own missing-input warning.
        p[0].Optional = true;
    }

    protected override void RegisterOutputParams(GH_OutputParamManager p)
    {
        p.AddCurveParameter("Segments", "S",
            "Surviving pieces, on the input path plus the index of the curve they came from. " +
            "A curve the gaps swallowed whole keeps its path and holds nothing, so later indices do not shift.",
            GH_ParamAccess.tree);
        p.AddIntegerParameter("Count", "N",
            "How many pieces each input curve produced, on the path it came in on.",
            GH_ParamAccess.tree);
    }

    protected override void SolveInstance(IGH_DataAccess da)
    {
        if (!NativeLibraryLoader.TryEnsureLoaded(out string loadError))
        {
            AddRuntimeMessage(GH_RuntimeMessageLevel.Error, loadError);
            return;
        }

        if (!da.GetDataTree(0, out GH_Structure<GH_Curve> curveTree) || curveTree.IsEmpty)
        {
            AddRuntimeMessage(GH_RuntimeMessageLevel.Remark, "No curves to shatter.");
            return;
        }

        double gapD = 0.0, touchTol = 0.0;
        bool testSelf = false;
        var plane = Plane.WorldXY;
        da.GetData(1, ref gapD);
        da.GetData(2, ref touchTol);
        da.GetData(3, ref testSelf);
        da.GetData(4, ref plane);

        if (gapD < 0.0)
        {
            AddRuntimeMessage(GH_RuntimeMessageLevel.Error, "Distance cannot be negative.");
            return;
        }

        if (touchTol < 0.0)
        {
            AddRuntimeMessage(GH_RuntimeMessageLevel.Error, "Tolerance cannot be negative.");
            return;
        }

        if (!plane.IsValid)
        {
            AddRuntimeMessage(GH_RuntimeMessageLevel.Error, "Plane is not valid.");
            return;
        }

        if (gapD == 0.0)
            AddRuntimeMessage(GH_RuntimeMessageLevel.Remark,
                "Distance is 0, so curves are cut at their contacts but no gap is opened.");

        var segmentsOut = new GH_Structure<GH_Curve>();
        var countsOut = new GH_Structure<GH_Integer>();

        for (int b = 0; b < curveTree.PathCount; b++)
        {
            GH_Path path = curveTree.Paths[b];
            IList<GH_Curve> branch = curveTree.Branches[b];

            // Refusing rather than approximating: a curve that is not a polyline has no
            // straight segments to meet, and reducing it to the line between its ends
            // would move every contact with nothing in the output to say so.
            var polylines = new List<Polyline>(branch.Count);
            var rejected = new List<int>();
            for (int i = 0; i < branch.Count; i++)
            {
                Curve? c = branch[i]?.Value;
                if (c == null || !c.TryGetPolyline(out Polyline pl) || pl.Count < 2)
                {
                    rejected.Add(i);
                    polylines.Add(new Polyline());
                }
                else
                {
                    polylines.Add(pl);
                }
            }

            if (rejected.Count > 0)
            {
                // Messages.Skipped is not reusable here: it ends "The rest were sorted",
                // and nothing is sorted or partially kept -- the branch is refused whole.
                string listed = string.Join(", ", rejected.Count > 20 ? rejected.GetRange(0, 20) : rejected);
                if (rejected.Count > 20)
                    listed += $", … ({rejected.Count - 20} more)";
                AddRuntimeMessage(GH_RuntimeMessageLevel.Error,
                    $"Branch {path}: {rejected.Count} of {branch.Count} curves are not polylines (index {listed}). " +
                    "Shatter Crossings handles polylines only, so nothing in this branch was computed.");
                segmentsOut.EnsurePath(path);
                countsOut.EnsurePath(path);
                continue;
            }

            // Explode into plane space. The contact test is 2D; working in plane
            // coordinates puts it on the plane the caller chose and leaves the height
            // above it as z, which the native side interpolates down each segment.
            int segCount = 0;
            foreach (Polyline pl in polylines)
                segCount += pl.Count - 1;

            if (segCount == 0)
            {
                segmentsOut.EnsurePath(path);
                countsOut.EnsurePath(path);
                continue;
            }

            var buffer = new double[segCount * 6];
            var owners = new int[segCount];
            var spans = new (int Start, int Stop)[polylines.Count];

            int s = 0;
            for (int owner = 0; owner < polylines.Count; owner++)
            {
                Polyline pl = polylines[owner];
                int start = s;
                for (int k = 0; k < pl.Count - 1; k++)
                {
                    Point3d a = ToPlane(plane, pl[k]);
                    Point3d z = ToPlane(plane, pl[k + 1]);
                    int o = s * 6;
                    buffer[o] = a.X; buffer[o + 1] = a.Y; buffer[o + 2] = a.Z;
                    buffer[o + 3] = z.X; buffer[o + 4] = z.Y; buffer[o + 5] = z.Z;
                    owners[s] = owner;
                    s++;
                }
                spans[owner] = (start, s);
            }

            GeomSeqCore.ShatterResult result;
            try
            {
                result = GeomSeqCore.ShatterAtCrossings(buffer, segCount, gapD, touchTol, owners, testSelf);
            }
            catch (Exception e) when (e is DllNotFoundException or EntryPointNotFoundException or BadImageFormatException)
            {
                AddRuntimeMessage(GH_RuntimeMessageLevel.Error, Messages.LoadFailedPrefix + e.Message);
                return;
            }

            // Where each input segment's pieces begin in the flat result.
            var offsets = new int[segCount + 1];
            for (int k = 0; k < segCount; k++)
                offsets[k + 1] = offsets[k] + result.PieceCounts[k];

            for (int owner = 0; owner < polylines.Count; owner++)
            {
                (int start, int stop) = spans[owner];
                GH_Path outPath = path.AppendElement(owner);

                // Pieces of one polyline, in order, stay contiguous wherever nothing was
                // cut away, so a run of pieces that touch end-to-start is one surviving
                // polyline. A gap inside a segment and a gap at a corner both break a run
                // by the same test, so neither needs a case of its own.
                var rebuilt = new List<GH_Curve>();
                var run = new List<Point3d>();

                for (int k = offsets[start]; k < offsets[stop]; k++)
                {
                    Line piece = result.Piece(k);
                    Point3d from = plane.PointAt(piece.FromX, piece.FromY, piece.FromZ);
                    Point3d to = plane.PointAt(piece.ToX, piece.ToY, piece.ToZ);

                    if (run.Count > 0 && run[run.Count - 1].DistanceTo(from) <= JoinTol)
                    {
                        run.Add(to);
                    }
                    else
                    {
                        if (run.Count > 1)
                            rebuilt.Add(new GH_Curve(new PolylineCurve(run)));
                        run = new List<Point3d> { from, to };
                    }
                }
                if (run.Count > 1)
                    rebuilt.Add(new GH_Curve(new PolylineCurve(run)));

                segmentsOut.AppendRange(rebuilt, outPath);
                countsOut.Append(new GH_Integer(rebuilt.Count), path);

                // An empty path rather than a missing one: dropping it would shift every
                // later index, which is exactly what this component promises not to do.
                if (rebuilt.Count == 0)
                    segmentsOut.EnsurePath(outPath);
            }
        }

        da.SetDataTree(0, segmentsOut);
        da.SetDataTree(1, countsOut);
    }

    /// <summary>World point to plane coordinates; x/y along the plane, z its height above it.</summary>
    private static Point3d ToPlane(Plane plane, Point3d world)
    {
        return plane.RemapToPlaneSpace(world, out Point3d local) ? local : world;
    }
}

File: src/gh/shatter_crossings_component.py
"""
Shatter Crossings Component for GeomSeq
GH entry point: cuts polylines where they cross and opens a gap at each crossing
(geometry_utils.shatter_at_crossings_native).

Polylines only. They are exploded into segments, shattered, and the pieces that
were never cut are rejoined -- so a 100-segment polyline crossed once comes back
as two polylines, not a hundred loose lines. Exploding is also what keeps the
door open for a sweep line later: it would replace how crossings are found,
without touching the explode/rejoin either side of it.
"""

import os
import sys

import Rhino
import Rhino.Geometry as rg
import ghpythonlib.treehelpers as th

try:
    import geomseq_core
except ImportError:
    # Dev mode: geomseq_core isn't installed / not on sys.path yet -- add its
    # parent src/ dir (this file lives in src/gh/, geomseq_core lives in src/geomseq_core/).
    _SRC = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    if _SRC not in sys.path:
        sys.path.insert(0, _SRC)

from geomseq_core import _reload

# Force-reload geometry_utils/misc so edits show up without restarting Rhino.
# Skip native_bridge -- reloading it would reload geomseq_core.dll every recompute (slow).
_reload.unload_modules("geomseq_core.misc")
_reload.unload_modules("geomseq_core.geometry_utils")

from geomseq_core.geometry_utils import shatter_at_crossings_native

# Pieces are built from the same doubles the vertices came from, so an uncut
# segment reproduces its endpoints exactly and a shared vertex is bit-identical.
# This only has to absorb the plane round-trip, which both sides of a joint go
# through together.
JOIN_TOL = 1e-9


class _Seg(object):
    """What curves_to_endpoint_buffer needs: two points, in plane space."""

    __slots__ = ("PointAtStart", "PointAtEnd")

    def __init__(self, a, b):
        self.PointAtStart, self.PointAtEnd = a, b


def _resolve_plane():
    """Unconnected optional GH input reads as world XY."""
    try:
        value = plane  # type: ignore
    except NameError:
        value = None
    return rg.Plane.WorldXY if value is None else value


def _resolve_gap_distance():
    """D is the whole gap: how far apart the two cut ends end up. Each crossing
    loses D/2 to either side of itself."""
    try:
        value = gap_distance  # type: ignore
    except NameError:
        value = None
    if value is None:
        print("[geomseq_core] D not connected -- using 0.0, so polylines are cut "
              "at crossings but no gap is opened")
        return 0.0
    return float(value)


def _resolve_touch_tolerance():
    """How close counts as touching rather than missing. A T-junction -- one
    polyline ending on another -- is a contact, and real ones are drawn a hair
    out, so the document tolerance is the right default rather than 0."""
    try:
        value = touch_tolerance  # type: ignore
    except NameError:
        value = None
    if value is not None:
        return float(value)
    try:
        return Rhino.RhinoDoc.ActiveDoc.ModelAbsoluteTolerance
    except Exception:
        return 0.0


def _resolve_self_intersection():
    try:
        value = self_intersection  # type: ignore
    except NameError:
        value = None
    return bool(value) if value is not None else False


def _to_polylines(curves):
    """(polylines, rejected) -- rejected holds the input indices that are not
    polylines. Arcs and general NURBS have no straight segments to cross, so
    they are refused rather than approximated by their control structure."""
    polylines, rejected = [], []
    for i, crv in enumerate(curves):
        if crv is None:
            rejected.append(i)
            continue
        ok, pl = crv.TryGetPolyline()
        if not ok or pl is None or pl.Count < 2:
            rejected.append(i)
        else:
            polylines.append((i, pl))
    return polylines, rejected


if __name__ == "__main__":
    src_plane = _resolve_plane()
    gap_d = _resolve_gap_distance()
    touch_tol = _resolve_touch_tolerance()
    test_self = _resolve_self_intersection()

    raw = list(curves) if curves else []  # type: ignore

    polylines, rejected = _to_polylines(raw)

    if rejected:
        # Refusing rather than warning: a non-polyline silently reduced to the
        # line between its ends would move the crossings, and nothing in the
        # output would say so.
        print("[geomseq_core] %d of %d input curves are not polylines (indices %s). "
              "Shatter Crossings handles polylines only -- nothing was computed."
              % (len(rejected), len(raw), rejected[:20]))
        segments_tree = th.list_to_tree([[]])
        piece_counts_tree = th.list_to_tree([[]])
    else:
        # --- explode, in plane space ---
        # The crossing test is 2D. Working in plane coordinates puts it on the
        # plane the caller chose, and the out-of-plane offset rides along as z:
        # the native side interpolates z down each segment, so a cut point keeps
        # the height it would have had.
        segs, owners, spans = [], [], []
        for owner, (src_index, pl) in enumerate(polylines):
            local = []
            for k in range(pl.Count):
                p = pl[k]
                ok, lp = src_plane.RemapToPlaneSpace(p)
                local.append(lp if ok else p)

            start = len(segs)
            for k in range(len(local) - 1):
                segs.append(_Seg(local[k], local[k + 1]))
                owners.append(owner)
            spans.append((src_index, start, len(segs)))

        shattered = shatter_at_crossings_native(segs, gap_d,
                                                touch_tol=touch_tol,
                                                segment_owner=owners,
                                                test_self=test_self)

        # --- rejoin, back in world space ---
        # A run of pieces that touch end-to-start is one surviving polyline. But
        # contiguity alone is not enough to decide that: at D = 0 the two sides
        # of a cut touch as well, and welding them back would make the whole
        # component look like a no-op. Where the piece came from settles it --
        # see below.
        nested_pieces, nested_counts = [], []
        for src_index, start, stop in spans:
            runs, cur = [], []
            for s_i in range(start, stop):
                for k, (s, e) in enumerate(shattered[s_i]):
                    # Only the FIRST piece of a segment can continue the run
                    # before it. Later pieces of the same segment follow a cut
                    # by construction, and at D = 0 a cut is contiguous -- so
                    # testing contiguity alone would silently weld them back
                    # together and the component would appear to do nothing.
                    joins = (k == 0 and cur and
                             all(abs(cur[-1][1][d] - s[d]) <= JOIN_TOL for d in (0, 1, 2)))
                    if joins:
                        cur.append((s, e))
                    else:
                        if cur:
                            runs.append(cur)
                        cur = [(s, e)]
            if cur:
                runs.append(cur)

            out_curves = []
            for run in runs:
                pts = [run[0][0]] + [e for _, e in run]
                world = [src_plane.PointAt(p[0], p[1], p[2]) for p in pts]
                out_curves.append(rg.PolylineCurve(world))

            nested_pieces.append(out_curves)
            nested_counts.append(len(out_curves))

        # Path {branch; input index}: a polyline cut in two stays addressable by
        # the index it went in with, and one the gaps swallowed whole keeps its
        # path and holds nothing rather than shifting every index after it.
        segments_tree = th.list_to_tree([nested_pieces])
        piece_counts_tree = th.list_to_tree([nested_counts])
